<a href="https://colab.research.google.com/github/kanyaweeud/data_viz/blob/main/Pandas102.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [3]:
import pandas as pd

In [4]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [5]:
import os

path = '/content/drive/My Drive/dataviz_2021_data'

covid_file_path = os.path.join(path,'pm-24-jan-2021.csv') 
print(covid_file_path)

/content/drive/My Drive/dataviz_2021_data/pm-24-jan-2021.csv


In [6]:
data_covid = pd.read_csv(covid_file_path)
data_covid.head()

No. announce_date  ... district_of_onset                              risk
0    1     1/12/2020  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
1    2     1/17/2020  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
2    3     1/22/2020  ...             เมือง  คนต่างชาติเดินทางมาจากต่างประเทศ
3    4     1/22/2020  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
4    5     1/24/2020  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ

[5 rows x 10 columns]

# .isnull()?? None??  

In [7]:
data_covid.isnull()

No.  announce_date  ...  district_of_onset   risk
0      False          False  ...               True  False
1      False          False  ...               True  False
2      False          False  ...              False  False
3      False          False  ...               True  False
4      False          False  ...               True  False
...      ...            ...  ...                ...    ...
13495  False          False  ...              False  False
13496  False          False  ...              False  False
13497  False          False  ...              False  False
13498  False          False  ...              False  False
13499  False          False  ...              False  False

[13500 rows x 10 columns]

.any() เอาค่าความจริงภายในแต่ละ column มา ```OR``` กัน  

In [8]:
data_covid.isnull().any()

No.                      False
announce_date            False
notification_date         True
sex                       True
age                       True
nationality               True
province_of_isolation     True
province_of_onset         True
district_of_onset         True
risk                      True
dtype: bool

.all() เอาค่าความจริงภายในแต่ละ column มา ```AND``` กัน  

In [9]:
data_covid.isnull().all()

No.                      False
announce_date            False
notification_date        False
sex                      False
age                      False
nationality              False
province_of_isolation    False
province_of_onset        False
district_of_onset        False
risk                     False
dtype: bool

In [10]:
data_covid.iloc[0,0].isnull()

AttributeError: ignored

In [11]:
data_covid['No.'][0].isnull()

AttributeError: ignored

In [12]:
data_covid.iloc[0,0]

1

In [13]:
data_covid.iloc[:1,0].isnull()

0    False
Name: No., dtype: bool

# ต่อตารางแกน X แกน y


*   ต่อแกน y คือ เพิ่ม records (เพิ่มจำนวนข้อมูล)
*   ต่อแกน x คือ เพิ่ม columns (เพิ่มรายละเอียดของข้อมูล)



## ต่อแกน Y pd.concat()

In [14]:
data_covid['province_of_onset']=='ขอนแก่น'

0        False
1        False
2        False
3        False
4        False
         ...  
13495    False
13496    False
13497    False
13498    False
13499    False
Name: province_of_onset, Length: 13500, dtype: bool

In [15]:
dataKK = data_covid[data_covid['province_of_onset']=='ขอนแก่น']
dataKK.head()

No.  ...                                           risk
180    181  ...                                        สนามมวย
462    463  ...    สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
1466  1467  ...                  คนไทยเดินทางกลับจากต่างประเทศ
1970  1971  ...  ไปสถานที่ชุมชน เช่น ตลาดนัด สถานที่ท่องเที่ยว
2637  2638  ...                  สัมผัสผู้เดินทางจากต่างประเทศ

[5 rows x 10 columns]

In [16]:
dataUD = data_covid[data_covid['province_of_onset']=='อุดรธานี']
dataUD.head()

No.  ...                                         risk
424  425  ...                                      สนามมวย
434  435  ...                                  สถานบันเทิง
471  472  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
883  884  ...                                  สถานบันเทิง
885  886  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้

[5 rows x 10 columns]

In [17]:
dataMS = data_covid[data_covid['province_of_onset']=='มหาสารคาม']
dataMS.head()

No.  ...                                         risk
346      347  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
789      790  ...                                  สถานบันเทิง
6690    6691  ...                            Cluster สมุทรสาคร
10802  10803  ...                            Cluster สมุทรสาคร

[4 rows x 10 columns]

In [18]:
dataMYisan = pd.concat([dataKK,dataUD,dataMS])
dataMYisan

No.  ...                                           risk
180      181  ...                                        สนามมวย
462      463  ...    สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
1466    1467  ...                  คนไทยเดินทางกลับจากต่างประเทศ
1970    1971  ...  ไปสถานที่ชุมชน เช่น ตลาดนัด สถานที่ท่องเที่ยว
2637    2638  ...                  สัมผัสผู้เดินทางจากต่างประเทศ
2673    2674  ...    สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
5948    5949  ...                              Cluster สมุทรสาคร
6082    6083  ...                              Cluster สมุทรสาคร
9333    9334  ...                           อยู่ระหว่างการสอบสวน
9334    9335  ...    สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
10610  10611  ...    สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
11517  11518  ...                           อยู่ระหว่างการสอบสวน
11697  11698  ...    สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
11698  11699  ...    สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
424      425  ...                                        สนามมวย
434      435  ...                                    สถานบันเทิง
471      472  ...    สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
883      884  ...                                    สถานบันเทิง
885      886  ...    สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
941      942  ...                                    สถานบันเทิง
1059    1060  ...                  คนไทยเดินทางกลับจากต่างประเทศ
2013    2014  ...                  คนไทยเดินทางกลับจากต่างประเทศ
5810    5811  ...                              Cluster สมุทรสาคร
7433    7434  ...                                  Cluster ระยอง
346      347  ...    สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
789      790  ...                                    สถานบันเทิง
6690    6691  ...                              Cluster สมุทรสาคร
10802  10803  ...                              Cluster สมุทรสาคร

[28 rows x 10 columns]

## ต่อแกน X 


*   จับ 2 ตารางมาต่อกันเลย (merge)
*   เลือกมาเพิ่มเฉพาะบาง column (map)



In [19]:
data_province = data_covid[['No.','announce_date','province_of_onset']]
data_province

No. announce_date province_of_onset
0          1     1/12/2020     กรุงเทพมหานคร
1          2     1/17/2020     กรุงเทพมหานคร
2          3     1/22/2020            นครปฐม
3          4     1/22/2020     กรุงเทพมหานคร
4          5     1/24/2020     กรุงเทพมหานคร
...      ...           ...               ...
13495  13496     1/24/2021     กรุงเทพมหานคร
13496  13497     1/24/2021     กรุงเทพมหานคร
13497  13498     1/24/2021     กรุงเทพมหานคร
13498  13499     1/24/2021     กรุงเทพมหานคร
13499  13500     1/24/2021     กรุงเทพมหานคร

[13500 rows x 3 columns]

In [20]:
data_human = data_covid[['No.','age','sex','nationality']]
data_human

No.   age   sex nationality
0          1  61.0  หญิง       China
1          2  74.0  หญิง       China
2          3  73.0  หญิง    Thailand
3          4  68.0   ชาย       China
4          5  66.0  หญิง       China
...      ...   ...   ...         ...
13495  13496  49.0  หญิง    Thailand
13496  13497  31.0   ชาย    Thailand
13497  13498  45.0   ชาย    Thailand
13498  13499  29.0  หญิง    Thailand
13499  13500  44.0  หญิง    Thailand

[13500 rows x 4 columns]

# แบบง่าย รู้ว่าสองตาราง record ตรงกัน

In [21]:
full_table1 = data_human.merge(data_province)
full_table1.head()

No.   age   sex nationality announce_date province_of_onset
0    1  61.0  หญิง       China     1/12/2020     กรุงเทพมหานคร
1    2  74.0  หญิง       China     1/17/2020     กรุงเทพมหานคร
2    3  73.0  หญิง    Thailand     1/22/2020            นครปฐม
3    4  68.0   ชาย       China     1/22/2020     กรุงเทพมหานคร
4    5  66.0  หญิง       China     1/24/2020     กรุงเทพมหานคร

## sort

In [22]:
data_human2 = data_human.sort_values('age')
data_human2

No.   age   sex nationality
1987    1988  0.10   ชาย    Thailand
11497  11498  0.11   ชาย       Burma
6477    6478  0.11  หญิง    Thailand
1675    1676  0.30   ชาย       Japan
1075    1076  0.40   ชาย    Thailand
...      ...   ...   ...         ...
13481  13482   NaN   NaN         NaN
13482  13483   NaN   NaN         NaN
13483  13484   NaN   NaN         NaN
13484  13485   NaN   NaN         NaN
13485  13486   NaN   NaN         NaN

[13500 rows x 4 columns]

In [23]:
full_table2 = data_human2.merge(data_province)
full_table2.head()

No.   age   sex nationality announce_date province_of_onset
0   1988  0.10   ชาย    Thailand      4/4/2020             ระยอง
1  11498  0.11   ชาย       Burma     1/16/2021         สมุทรสาคร
2   6478  0.11  หญิง    Thailand    12/30/2020          เพชรบุรี
3   1676  0.30   ชาย       Japan      4/1/2020     กรุงเทพมหานคร
4   1076  0.40   ชาย    Thailand     3/27/2020      สุราษฎร์ธานี

In [24]:
data_human2_renamed = data_human2.rename(columns={'No.': 'patientNumber'})
data_human2_renamed.head()

patientNumber   age   sex nationality
1987            1988  0.10   ชาย    Thailand
11497          11498  0.11   ชาย       Burma
6477            6478  0.11  หญิง    Thailand
1675            1676  0.30   ชาย       Japan
1075            1076  0.40   ชาย    Thailand

In [25]:
data_human2_renamed.merge(data_province)

MergeError: ignored

In [36]:
full_table3 = data_human2_renamed.merge(data_province,left_on='patientNumber',right_on='No.')
full_table3.head()

patientNumber   age   sex  ...    No.  announce_date province_of_onset
0           1988  0.10   ชาย  ...   1988       4/4/2020             ระยอง
1          11498  0.11   ชาย  ...  11498      1/16/2021         สมุทรสาคร
2           6478  0.11  หญิง  ...   6478     12/30/2020          เพชรบุรี
3           1676  0.30   ชาย  ...   1676       4/1/2020     กรุงเทพมหานคร
4           1076  0.40   ชาย  ...   1076      3/27/2020      สุราษฎร์ธานี

[5 rows x 9 columns]

## map() เลือกมาเฉพาะบาง column มาแปะเพิ่มเข้าไป

In [26]:
data_human2_renamed.head()

patientNumber   age   sex nationality
1987            1988  0.10   ชาย    Thailand
11497          11498  0.11   ชาย       Burma
6477            6478  0.11  หญิง    Thailand
1675            1676  0.30   ชาย       Japan
1075            1076  0.40   ชาย    Thailand

In [29]:
data_covid.head()

No. announce_date  ... district_of_onset                              risk
0    1     1/12/2020  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
1    2     1/17/2020  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
2    3     1/22/2020  ...             เมือง  คนต่างชาติเดินทางมาจากต่างประเทศ
3    4     1/22/2020  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
4    5     1/24/2020  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ

[5 rows x 10 columns]

คุณสมบัติของ pandas ก็คือเราสามารถสร้าง column ใหม่ให้ตาราง df ได้ โดย 

```df['ชื่อ column ใหม่'] = (list ที่มีจำนวนสมาชิกเท่ากับจำนวน record ของ df)```

In [27]:
data_human2_renamed.head()

patientNumber   age   sex nationality
1987            1988  0.10   ชาย    Thailand
11497          11498  0.11   ชาย       Burma
6477            6478  0.11  หญิง    Thailand
1675            1676  0.30   ชาย       Japan
1075            1076  0.40   ชาย    Thailand

In [28]:
data_human2_renamed['num'] = range(data_human2_renamed.shape[0])
data_human2_renamed

patientNumber   age   sex nationality    num
1987            1988  0.10   ชาย    Thailand      0
11497          11498  0.11   ชาย       Burma      1
6477            6478  0.11  หญิง    Thailand      2
1675            1676  0.30   ชาย       Japan      3
1075            1076  0.40   ชาย    Thailand      4
...              ...   ...   ...         ...    ...
13481          13482   NaN   NaN         NaN  13495
13482          13483   NaN   NaN         NaN  13496
13483          13484   NaN   NaN         NaN  13497
13484          13485   NaN   NaN         NaN  13498
13485          13486   NaN   NaN         NaN  13499

[13500 rows x 5 columns]

In [30]:
data_human2_renamed['patientNumber'].map(data_covid.set_index('No.')['risk']) #.map มันจะ ผูกค่า 'patientNumber' กับ 'No.' เข้าด้วยกัน แล้ว return list ของค่า ['risk']

1987     สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
11497                              Cluster สมุทรสาคร
6477     สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
1675     สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
1075     สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
                            ...                     
13481                              Cluster สมุทรสาคร
13482                              Cluster สมุทรสาคร
13483                              Cluster สมุทรสาคร
13484                              Cluster สมุทรสาคร
13485                              Cluster สมุทรสาคร
Name: patientNumber, Length: 13500, dtype: object

In [31]:
data_human2_renamed['detail'] = data_human2_renamed['patientNumber'].map(data_covid.set_index('No.')['risk'])
data_human2_renamed

patientNumber   age  ...    num                                       detail
1987            1988  0.10  ...      0  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
11497          11498  0.11  ...      1                            Cluster สมุทรสาคร
6477            6478  0.11  ...      2  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
1675            1676  0.30  ...      3  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
1075            1076  0.40  ...      4  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
...              ...   ...  ...    ...                                          ...
13481          13482   NaN  ...  13495                            Cluster สมุทรสาคร
13482          13483   NaN  ...  13496                            Cluster สมุทรสาคร
13483          13484   NaN  ...  13497                            Cluster สมุทรสาคร
13484          13485   NaN  ...  13498                            Cluster สมุทรสาคร
13485          13486   NaN  ...  13499                            Cluster สมุทรสาคร

[13500 rows x 6 columns]

## HW7 สร้างตารางใหม่ ที่ค่าใน ```sex``` เป็น missing ทั้งหมด  
- สรุปว่าทำไม record นั้นๆถึงเป็น missing

In [39]:
new_table = data_covid.fillna(value={'sex':'missing'})
new_table.head()

No. announce_date  ... district_of_onset                              risk
0    1     1/12/2020  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
1    2     1/17/2020  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
2    3     1/22/2020  ...             เมือง  คนต่างชาติเดินทางมาจากต่างประเทศ
3    4     1/22/2020  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
4    5     1/24/2020  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ

[5 rows x 10 columns]

In [42]:
missing = new_table[new_table['sex']=='missing']
missing

No. announce_date  ... district_of_onset               risk
4391    4392    12/20/2020  ...             เมือง  Cluster สมุทรสาคร
4392    4393    12/20/2020  ...             เมือง  Cluster สมุทรสาคร
4393    4394    12/20/2020  ...             เมือง  Cluster สมุทรสาคร
4394    4395    12/20/2020  ...             เมือง  Cluster สมุทรสาคร
4395    4396    12/20/2020  ...             เมือง  Cluster สมุทรสาคร
...      ...           ...  ...               ...                ...
13481  13482     1/24/2021  ...               NaN  Cluster สมุทรสาคร
13482  13483     1/24/2021  ...               NaN  Cluster สมุทรสาคร
13483  13484     1/24/2021  ...               NaN  Cluster สมุทรสาคร
13484  13485     1/24/2021  ...               NaN  Cluster สมุทรสาคร
13485  13486     1/24/2021  ...               NaN  Cluster สมุทรสาคร

[2924 rows x 10 columns]

In [44]:
missing[missing['province_of_onset']=='สมุทรสาคร'].shape 

(1779, 10)

In [52]:
data_covid[data_covid['province_of_onset']=='สมุทรสาคร'].shape 

(3904, 10)

In [69]:
missing[missing['province_of_onset']=='ขอนแก่น'].shape 

(0, 10)

In [68]:
missing[missing['province_of_onset']=='กรุงเทพ'].shape 

(0, 10)

In [62]:
missing[missing['nationality']=='Burma'].shape France

(1393, 10)

In [65]:
missing[missing['nationality']=='Thailand'].shape

(54, 10)

สรุป 
- จากการสร้างตารางใหม่พบว่าค่าใน sex ที่มี missing มีจำนวน 2,924 คน
- ผู้ติดเชื้อในสมุทรสาครที่ไม่ระบุเพศมีทั้งหมด 1,779 คน จากทั้งหมด 3,904
- จังหวัดอื่นๆ ไม่พบ missing ในคอลัมน์ sex เช่น ขอนแก่น กรุงเทพ กระบี่ เชี่ยงใหม่ เป็นต้น
- การที่มี missing อาจจะเป็นเพราะมียอดผู้ติดเชื้อจำนวนมากในแต่ละวัน และทั้งเป็นคนเชื้อชาติอื่นที่เข้ามาทำงานในไทยเป็นจำนวนมาก ทำให้ไม่สามารถเก็บข้อมูลเพศได้หมด